Tpc worker pool

dynamic-entropy opened the pull request:
Fix https://github.com/xrootd/xrootd/issues/2401

bbockelm committed:
[XrdHttpTpc] Framework for worker pool for TPC requests

Co-authored-by: Rahul Chauhan <rahul.chauhan@cern.ch>
Signed-off-by: Rahul Chauhan <rahul.chauhan@cern.ch>

dynamic-entropy committed:
Updated perform_http_tpc to correctly detect
transfer success or failure by checking the last line
of the chunked reponse, since the HTTP 201 code
is always returned regardless of actual status.

Signed-off-by: Rahul Chauhan <rahul.chauhan@cern.ch>

dynamic-entropy committed:
[XrdHttpTpc] Change ownership of curl socket callbacks

Signed-off-by: Rahul Chauhan <rahul.chauhan@cern.ch>

dynamic-entropy committed:
[XrdHttpTpc] Add scitag flow configuration to test
Add scitag flow query string in tests

Signed-off-by: Rahul Chauhan <rahul.chauhan@cern.ch>

dynamic-entropy committed:
[XrdHttpTpc] Refactor PMarkManager and TPCRequestManager to add SciTag integration

Signed-off-by: Rahul Chauhan <rahul.chauhan@cern.ch>

bbockelm commented:
@dynamic-entropy and I discussed this PR a bit over Zoom.  Beyond the code review itself, items we want to work on:
- Get this PR into a patch build and deploy at a site that actively does packet marking to look for regressions "at scale".
- Implement the `Cancel()` method and write a unit test to ensure it works.  (The unit test should start a COPY, disconnect, and look to ensure the transfer does not continue to completion).
- We don't recall whether a slow/stalled transfer would ever automatically shutdown (or if this is always triggered by FTS).  Test with and without this patch, by hand, to make sure there is no change in behavior (we might desire a change in behavior later but that should be a separate PR).
- Look at the performance markers and ensure the information is the same before & after this patch.  We don't think there's any unit test coverage for the performance marker contents.  Particularly, does getting the [remote connection list](https://github.com/xrootd/xrootd/blob/346858c2ce7813e20c7d10b17a4ac583469ba200/src/XrdHttpTpc/XrdHttpTpcTPC.cc#L425) still work?

dynamic-entropy committed:
[XrdHttpTpc] Add test for cancellations

Signed-off-by: Rahul Chauhan <rahul.chauhan@cern.ch>

dynamic-entropy committed:
[XrdHttpTpc] Implement cancellation in TPC Pool

Signed-off-by: Rahul Chauhan <rahul.chauhan@cern.ch>

dynamic-entropy committed:
[XrdHttpTpc] Add multistream transfers to the test

Signed-off-by: Rahul Chauhan <rahul.chauhan@cern.ch>

dynamic-entropy committed:
[XrdHttpTpc] Update remote connection list from curl

Signed-off-by: Rahul Chauhan <rahul.chauhan@cern.ch>

bbockelm commented on src/XrdHttpTpc/XrdHttpTpcPool.cc:
```diff
@@ -0,0 +1,479 @@
+	time_t last_update = now - update_interval; // Inorder to always fetch on first pass 
+
+	auto fail_and_exit = [&](int code, const std::string& msg) -> bool {
+      curl_multi_remove_handle(multi_handle, curl);
```
Fix the inconsistent whitespace mapping here.

bbockelm commented on src/XrdHttpTpc/XrdHttpTpcPool.cc:
```diff
@@ -0,0 +1,479 @@
+            if (!iter->second->IsDone()){
+            	queue = iter->second;
+                break;
+			}
```
Another small whitespace issue.

bbockelm commented on src/XrdHttpTpc/XrdHttpTpcTPC.cc:
```diff
@@ -1004,7 +843,7 @@ int TPCHandler::ProcessPullReq(const std::string &resource, XrdHttpExtReq &req)
     state.SetupHeaders(req);
     state.SetContentLength(sourceFileContentLength);
-
+	
```
avoid unnecessary whitespace changes.

bbockelm commented on src/XrdHttpTpc/XrdHttpTpcTPC.hh:
```diff
@@ -117,6 +112,7 @@ private:
     int SendPerfMarker(XrdHttpExtReq &req, TPCLogRecord &rec, TPC::State &state);
     int SendPerfMarker(XrdHttpExtReq &req, TPCLogRecord &rec, std::vector<State*> &state,
         off_t bytes_transferred);
+	int SendPerfMarker(XrdHttpExtReq &req, TPCLogRecord &rec, TPC::State &state, std::string desc);
```
Fix whitespace alignment.

bbockelm reviewed (commented):
Getting quite close!

- Please squash a few of the debugging commits.
- Review through the code and make sure no unnecessary developer debugging messages have snuck through.
- Some errant whitespace issues have crept up (lines not aligning, switching between tab & spaces).  Try setting your IDE to have tab stop be at 8 spaces to make these accidental changes more visible as you're fixing them.
- It's great that we can now limit the number of concurrent transfers!  Let's make these configurable, `tpc.max_active_transfers_per_queue` and `tpc.max_waiting_transfers_per_queue`?  Let's set the `max_active` to 50 by default and keep `max_waiting` where it is.
- For selection of queues, we should include the experiment name.  This way, if no scitag is enabled, we still partition between different experiments.

dynamic-entropy reviewed (commented):
Squashed the debugging commits and cleaned some changes. Still kept some whitespace changes - they correct a previous mistake of mixing the use of tabs and spaces for indentation in the test bash scripts.

dynamic-entropy committed:
[XrdHttpTpc] Make max_transfers and max_pending requests configurable

Signed-off-by: Rahul Chauhan <rahul.chauhan@cern.ch>

bbockelm commented on src/XrdHttpTpc/XrdHttpTpcConfigure.cc:
```diff
@@ -96,6 +96,22 @@ bool TPCHandler::Configure(const char *configfn, XrdOucEnv *myEnv)
+                m_log.Emsg("Config", "tpc.max_active_transfers_per_queue value not specified.");
+                return false;
+            }
+            unsigned int max_active = std::stoi(val);
```
I think this can switch to `XrdOuca2x::a2i`.  It's undocumented but here's the signature:

```
int XrdOuca2x::a2i(XrdSysError &Eroute, const char *emsg, const char *item,
                                             int *val, int minv, int maxv)
```

The `minv` and `maxv` arguments are "minimum value" and "maximum value" which is likely useful here.  It also takes care of logging of error messages, which may shorten the code a bit.

dynamic-entropy committed:
[XrdHttpTpc] Use vorg as part of the queue identifier as well as worker and request labels

Also fix nomenclature of identifier and labels:
A queue has a identifier (unique)
Workers and Requets have labels (not uniquet) that map to a queue

Signed-off-by: Rahul Chauhan <rahul.chauhan@cern.ch>

bbockelm commented:
FYI - this has been running at the CMS T2 in Nebraska for a little under 2 weeks, including a mini-DC.  No issues were reported by the Nebraska team.

amadio commented on src/XrdHttpTpc/XrdHttpTpcTPC.cc:
```diff
@@ -407,11 +407,30 @@ int TPCHandler::GetContentLengthTPCPull(CURL *curl, XrdHttpExtReq &req, uint64_t
+    rec.bytes_transferred = state.BytesTransferred();
+    logTransferEvent(LogMask::Debug, rec, "PERF_MARKER");
+    return req.ChunkResp(ss.str().c_str(), 0);
+}
```
This is adding code duplication, we already have two `TPCHandler::SendPerfMarker` methods. We should not add a third. I think it's better to think how to consolidate them into a single one instead.

Also, this commit is mixing up a lot more changes than what it says it's doing (`Update remote connection list from curl`). Please try to structure changes as one logical change per commit.

amadio reviewed (changes requested):
I'd like to understand better the motivation for tying the worker pools to scitags and vorgs, as these are not always going to be available. It's not clear to me that this strategy will get us better results. I think we should look at real world cases to have a more informed decision on how to group transfers into the worker pools.

amadio commented:
@abh3 Please let us know your thoughts here too. Thank you!

abh3 reviewed (commented):
I have also discussed this with @amadio and I too am mystified on why the queues are based on scitags. It definitely complicates things doing it that way. I think we really need to have another discussion on what the ultimate goal is and consider simpler ways of achieving it.

amadio commented:
Due to the problems reported in the users list linked above, merge conflicts, and lack of consensus, I'm closing this for now. Let's discuss this in our development meetings to have a fresh replacement which is not tied to scitags and that is based on XRootD 6.0.
